libraries/standard/http/test/test-HTTPClient_ReadHeader.c

```diff
@@ -0,0 +1,228 @@
+#include <string.h>
+#include <stdbool.h>
+
+#include "common.h"
+
+/* Mirror of internal type used by HTTPClient_ReadHeader function. */
+typedef struct readHeaderContext
+{
+    const uint8_t * pHeaderName;
+    size_t headerNameLen;
+    const uint8_t ** pHeaderValueLoc;
+    size_t * pHeaderValueLen;
+    uint8_t headerFound;
+} readHeaderContext_t;
+
+/* Functions are pulled out into their own C files to be tested as a unit. */
+#include "readHeaderParsingCallback.c"
+#include "HTTPClient_ReadHeader.c"
+
+/* Template HTTP request for a PUT request. */
+static const char * pTestResponse = "HTTP/1.1 200 OK"
+                                    "test-header0: test-value0\r\n"
+                                    "test-header1: test-value1\r\n"
+                                    "test-header2: test-value2\r\n"
+                                    "\r\n";
+static const size_t headerFieldInRespLoc = 15;
+static const size_t headerFieldInRespLen = strlen( "test-header0" );
+static const size_t headerValInRespLoc = 29;
+static const size_t headerValInRespLen = strlen( "test-value0" );
+
+#define HEADER_IN_BUFFER        "test-header0"
+#define HEADER_NOT_IN_BUFFER    "header-not-in-buffer"
+
+/* Mocked out implementations of parser function dependencies. */
+static HTTPStatus_t initializeParsingContextRetCode = HTTP_SUCCESS;
+HTTPStatus_t HTTPClient_InitializeParsingContext( HTTPParsingContext_t * pParsingContext,
```

Comment:
This function is private despite it being in a different file, and I don't think it should be mocked. All the cases should be reached from the public functions.

Reply:
From the perspective of the design concept of "Separation of Concerns", keeping the API and Parser modules should be de-coupled from each other enables the code's implementation to be non-rigid or flexible. (In case, the underlying parser is changed, the API module code stays unchanged with changing the parser).
Similarly, unit tests should be scoped only to the specific "module" (or function) being tested with all its (external) dependencies mocked out, so that the function being tested can be comprehensively tested by tweaking the inputs from all sides into the function

Reply:
I agree that unit tests should be written for public functions and every statement of all the internal/static functions must be reachable via public functions by simulating all the possible scenarios by mocking external dependencies. If there are statements in internal functions which are not reachable via public APIs, that is dead code anyways. In addition, I am not sure it will be possible to test internal functions with cmock.

Reply:
Yah don't think it's possible to test internal functions with cmock because it takes in a header file.

Reply:
Here, the parser function is not an internal function, but a function exposed by the `http_client_parse.h` file.

Reply:
When we test we want to isolate as much as possible. The smaller the set we are testing the easier it is to identify the problem. Regardless of the public interface we still want to unit test as finely grained as we can. In this case it seems possible and easy to mock the other file and test it separetely. 

My advice would be to test every C file by itself, when there are 2 files then always mock the one when you test the other.

Reply:
> I am not sure it will be possible to test internal functions with cmock.

If the function has it's own header file and c file it is always possible to test it using cmock. The way CMock works in this regard is basically just how the C standard works. If you have a compilation unit with an external symbol you can always link it into a program, the unit test for that compilation unit will ensure that the compilation unit works as expected.

Even if code is not public-facing, if it is split into a separate module we imply that it is separated because of separation of conerns principle, this principle will apply to the unit tests in all the same ways for all the same reasons. 

What I am saying is that if you can make an argument that the unit cannot stand alone then you are making an argument that it should not be a unit at all and the code should be placed in the C file it is being used in, because then it really is part of that C file and the functions should be made static. If that does not make sense it needs to be separated during testing and have its own unit tests ...

Reply:
I agree with what @cobusve mentioned. What I meant by "internal" is `static functions` which are not exported out of a compilation unit and what I meant by "public" functions is all `non-static` functions which are exported out of a compilation unit. Seems like we are in agreement that we need to unit test all the `non-static` functions and not unit test `static` functions.

Thanks.

Reply:
Per offline discussion, we have decided not to put the http-parser wrapped functions in their files not only because they cannot stand-alone as their own unit, but also since parsing is the backbone of the HTTPClient library it is in the same "area of concern" . The parsing functions will be made static.